### PR TITLE
Fix statistics

### LINKED ISSUE #2790: pdal info incorrectly computes the variance and standard deviation for some attributes
**Describe the bug**
PDAL incorrectly computes the variance and std dev on some attributes. In this example, pdal info shows a standard deviation of 0.306 and a variance of 0.093 on an attribute with a minimum *and* maximum of 1000. 

To reproduce
1. Start with a LAZ file (I used autzen)
2. Create a pipeline that assigns all Point Source ID's to 1000
3. Run `pdal info --stats` on the output file and look at the standard deviation and variance of the Point Source ID's

<details>
  
```
$ cat pipeline.json
{
    "pipeline": [
        "autzen-in.laz",
        { "type": "filters.assign", "assignment": "PointSourceId[0:]=1000"},
        { "type": "writers.las", "filename": "autzen-out.laz", "compression": "laszip" }
    ]
}
$ pdal pipeline pipeline.json
$ pdal info --stats autzen-out.laz
 {
  "filename": "autzen-out.laz",
  "pdal_version": "2.0.1 (git-version: 3449cd)",
  "stats":
  {
    "bbox":
    {
      "EPSG:4326":
      {
        "bbox":
        {
          "maxx": -123.0619599,
          "maxy": 44.06278031,
          "maxz": 187.531248,
          "minx": -123.0755422,
          "miny": 44.04971882,
          "minz": 123.791472
        },
        "boundary": "{ \"type\": \"Polygon\", \"coordinates\": [ [ [ -123.074986743550582, 44.04971881880158, 123.791472 ], [ -123.075542228546169, 44.062486228152146, 123.791472 ], [ -123.062512596996228, 44.062780307016681, 187.531248 ], [ -123.061959918532395, 44.050012834694726, 187.531248 ], [ -123.074986743550582, 44.04971881880158, 123.791472 ] ] ] }"
      },
      "native":
      {
        "bbox":
        {
          "maxx": 639003.73,
          "maxy": 853537.66,
          "maxz": 615.26,
          "minx": 635577.79,
          "miny": 848882.15,
          "minz": 406.14
        },
        "boundary": "{ \"type\": \"Polygon\", \"coordinates\": [ [ [ 635577.79, 848882.15, 406.14 ], [ 635577.79, 853537.66, 406.14 ], [ 639003.73, 853537.66, 615.26 ], [ 639003.73, 848882.15, 615.26 ], [ 635577.79, 848882.15, 406.14 ] ] ] }"
      }
    },
    "statistic":
    [
      {
        "average": 637294.1783,
        "count": 10653336,
        "maximum": 639003.73,
        "minimum": 635577.79,
        "name": "X",
        "position": 0,
        "stddev": 967.9329805,
        "variance": 936894.2548
      },
      {
        "average": 851247.6953,
        "count": 10653336,
        "maximum": 853537.66,
        "minimum": 848882.15,
        "name": "Y",
        "position": 1,
        "stddev": 1322.356387,
        "variance": 1748626.415
      },
      {
        "average": 434.1025002,
        "count": 10653336,
        "maximum": 615.26,
        "minimum": 406.14,
        "name": "Z",
        "position": 2,
        "stddev": 24.67893148,
        "variance": 609.0496589
      },
      {
        "average": 77.14742312,
        "count": 10653336,
        "maximum": 254,
        "minimum": 0,
        "name": "Intensity",
        "position": 3,
        "stddev": 62.62422344,
        "variance": 3921.793362
      },
      {
        "average": 1.17801438,
        "count": 10653336,
        "maximum": 4,
        "minimum": 1,
        "name": "ReturnNumber",
        "position": 4,
        "stddev": 0.4653418642,
        "variance": 0.2165430505
      },
      {
        "average": 1.358579791,
        "count": 10653336,
        "maximum": 4,
        "minimum": 1,
        "name": "NumberOfReturns",
        "position": 5,
        "stddev": 0.6656066447,
        "variance": 0.4430322055
      },
      {
        "average": 0.4989654884,
        "count": 10653336,
        "maximum": 1,
        "minimum": 0,
        "name": "ScanDirectionFlag",
        "position": 6,
        "stddev": 0.4999993213,
        "variance": 0.2499993213
      },
      {
        "average": 0,
        "count": 10653336,
        "maximum": 0,
        "minimum": 0,
        "name": "EdgeOfFlightLine",
        "position": 7,
        "stddev": 0,
        "variance": 0
      },
      {
        "average": 1.256686262,
        "count": 10653336,
        "maximum": 2,
        "minimum": 1,
        "name": "Classification",
        "position": 8,
        "stddev": 0.436805292,
        "variance": 0.1907988632
      },
      {
        "average": -0.812061405,
        "count": 10653336,
        "maximum": 20,
        "minimum": -21,
        "name": "ScanAngleRank",
        "position": 9,
        "stddev": 8.484319324,
        "variance": 71.98367439
      },
      {
        "average": 126.4052859,
        "count": 10653336,
        "maximum": 156,
        "minimum": 115,
        "name": "UserData",
        "position": 10,
        "stddev": 3.833000243,
        "variance": 14.69189086
      },
      {
        "average": 1000,
        "count": 10653336,
        "maximum": 1000,
        "minimum": 1000,
        "name": "PointSourceId",
        "position": 11,
        "stddev": 0.3063777399,
        "variance": 0.09386731948
      },
      {
        "average": 121.3214254,
        "count": 10653336,
        "maximum": 255,
        "minimum": 35,
        "name": "Red",
        "position": 12,
        "stddev": 45.56263834,
        "variance": 2075.954013
      },
      {
        "average": 126.2526972,
        "count": 10653336,
        "maximum": 255,
        "minimum": 49,
        "name": "Green",
        "position": 13,
        "stddev": 36.85451838,
        "variance": 1358.255525
      },
      {
        "average": 111.2207554,
        "count": 10653336,
        "maximum": 255,
        "minimum": 49,
        "name": "Blue",
        "position": 14,
        "stddev": 31.95561927,
        "variance": 1021.161603
      },
      {
        "average": 247608.4011,
        "count": 10653336,
        "maximum": 249783.703,
        "minimum": 245369.8966,
        "name": "GpsTime",
        "position": 15,
        "stddev": 1178.538827,
        "variance": 1388953.768
      }
    ]
  }
}


```

</details>

**Expected behavior**
I would expect that an attribute with a minimum, maximum and average of N would have 0 variance and 0 standard deviation

**System/installation information:**

```
$ uname -a
Linux ryan-XPS-15-9560 5.0.0-32-generic #34~18.04.2-Ubuntu SMP Thu Oct 10 10:36:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ pdal --version
```

My pdal is installed via conda.

<details>
  
```
$ conda list
# packages in environment at /home/rj/miniconda3/envs/pdal:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
boost-cpp                 1.70.0               h8e57a91_2    conda-forge
bzip2                     1.0.8                h7b6447c_0  
ca-certificates           2019.10.16                    0  
cairo                     1.16.0            hfb77d84_1002    conda-forge
certifi                   2019.9.11                py37_0  
cfitsio                   3.470                hb60a0a2_2    conda-forge
curl                      7.65.3               hbc83047_0  
cython                    0.29.13                  pypi_0    pypi
expat                     2.2.6                he6710b0_0  
fontconfig                2.13.1            h86ecdb6_1001    conda-forge
freetype                  2.9.1                h8a8886c_1  
freexl                    1.0.5                h14c3975_0  
gdal                      3.0.1            py37hbb6b9fb_8    conda-forge
geos                      3.7.2                he1b5a44_1    conda-forge
geotiff                   1.5.1                hcd53e25_3    conda-forge
gettext                   0.19.8.1             hd7bead4_3  
giflib                    5.1.7                h516909a_1    conda-forge
glib                      2.58.3            h6f030ca_1002    conda-forge
hdf4                      4.2.13               h3ca952b_2  
hdf5                      1.10.5          nompi_h3c11f04_1103    conda-forge
icu                       64.2                 he1b5a44_1    conda-forge
jpeg                      9c                h14c3975_1001    conda-forge
json-c                    0.13.1               h1bed415_0  
jsoncpp                   1.8.4                hfd86e86_0  
kealib                    1.4.10            h58c409b_1005    conda-forge
krb5                      1.16.3            h05b26f9_1001    conda-forge
laszip                    3.4.1                he1b5a44_0    conda-forge
laz-perf                  1.3.0             he1b5a44_1002    conda-forge
libblas                   3.8.0               12_openblas    conda-forge
libcblas                  3.8.0               12_openblas    conda-forge
libcurl                   7.65.3               h20c2e04_0  
libdap4                   3.20.2            hd48c02d_1000    conda-forge
libedit                   3.1.20181209         hc058e9b_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 9.1.0                hdf63c60_0  
libgdal                   3.0.1                hf47eb90_8    conda-forge
libgfortran-ng            7.3.0                hdf63c60_0  
libiconv                  1.15                 h63c8f33_5  
libkml                    1.3.0             h4fcabce_1010    conda-forge
liblapack                 3.8.0               12_openblas    conda-forge
libnetcdf                 4.6.2             h303dfb8_1003    conda-forge
libopenblas               0.3.7                h6e990d7_1    conda-forge
libpng                    1.6.37               hbc83047_0  
libpq                     11.5                 hd9ab2ff_1    conda-forge
libspatialite             4.3.0a            h57ae47a_1030    conda-forge
libssh2                   1.8.2                h1ba5d50_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.0.10            h57b8799_1003    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libxcb                    1.13                 h1bed415_1  
libxml2                   2.9.9                hea5a465_1  
lz4-c                     1.8.3             he1b5a44_1001    conda-forge
ncurses                   6.1                  he6710b0_1  
nitro                     2.7.dev5          he1b5a44_1002    conda-forge
nodejs                    11.11.0              hf484d3e_0    conda-forge
numpy                     1.17.2           py37h95a1406_0    conda-forge
openjpeg                  2.3.1                h21c5421_1    conda-forge
openssl                   1.1.1d               h7b6447c_3  
packaging                 19.2                       py_0  
pcre                      8.43                 he6710b0_0  
pdal                      2.0.1            py37h7e15382_5    conda-forge
pip                       19.3.1                   py37_0  
pixman                    0.38.0               h7b6447c_0  
poppler                   0.67.0               ha967d66_7    conda-forge
poppler-data              0.4.9                         0  
postgresql                11.5                 hc63931a_1    conda-forge
proj4                     6.1.1                hc80f0dc_1    conda-forge
pyparsing                 2.4.2                      py_0  
python                    3.7.3                h33d41f4_1    conda-forge
python-pdal               2.2.0            py37hc11b287_0    conda-forge
readline                  8.0                  hf8c457e_0    conda-forge
setuptools                41.4.0                   py37_0  
six                       1.12.0                   py37_0  
sqlite                    3.30.1               h7b6447c_0  
tbb                       2018.0.5             h6bb024c_0  
tiledb                    1.6.2                h69c774e_1    conda-forge
tk                        8.6.9             hed695b0_1002    conda-forge
tzcode                    2019a             h516909a_1002    conda-forge
wheel                     0.33.6                   py37_0  
xerces-c                  3.2.2             h8412b87_1004    conda-forge
xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
xorg-libice               1.0.10               h516909a_0    conda-forge
xorg-libsm                1.2.3             h84519dc_1000    conda-forge
xorg-libx11               1.6.8                h516909a_0    conda-forge
xorg-libxext              1.3.4                h516909a_0    conda-forge
xorg-libxrender           0.9.10            h516909a_1002    conda-forge
xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
xorg-xproto               7.0.31            h14c3975_1007    conda-forge
xz                        5.2.4                h14c3975_4  
zlib                      1.2.11               h7b6447c_3  
zstd                      1.4.0                h3b9ef0a_0    conda-forge

```

</details>

<details>
  
```
$ conda info

     active environment : pdal
    active env location : /home/rj/miniconda3/envs/pdal
            shell level : 2
       user config file : /home/rj/.condarc
 populated config files : 
          conda version : 4.7.11
    conda-build version : not installed
         python version : 3.7.4.final.0
       virtual packages : __cuda=9.1
       base environment : /home/rj/miniconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/rj/miniconda3/pkgs
                          /home/rj/.conda/pkgs
       envs directories : /home/rj/miniconda3/envs
                          /home/rj/.conda/envs
               platform : linux-64
             user-agent : conda/4.7.11 requests/2.22.0 CPython/3.7.4 Linux/5.0.0-32-generic ubuntu/18.04.3 glibc/2.27
                UID:GID : 1001:1001
             netrc file : None
           offline mode : False


```

</details>

### PR DESCRIPTION
**Pull request by abellgithub**
Close #2790

### CODE REVIEW

File: filters/StatsFilter.hpp [hunk 1]
```diff
@@ -68,15 +68,52 @@ typedef std::vector<double> DataVector;
...
+        if (M2 == 0 || !m_advanced)
+            return 0;
+        return kurtosis() - 3;
+    }
+    double sampleKurtosis() const
+    {
+        if (M2 == 0 || m_cnt <= 3 || !m_advanced)
+            return 0;
+        double c(m_cnt);
+        return kurtosis() * (c + 1) * (c - 1) / ((c - 2) * (c - 3));
```
Comment (chambbj) [inline filters/StatsFilter.hpp]: I didn't go through the math, so I'll just ask, was it intentional to do `(c+1)*(c-1)`? I see that it is `(c-1)*(c-1)` in the sample excess kurtosis, so I'm guessing that once you expand all the terms there must be some cancellation that is going on.

Comment (abellgithub) [reply filters/StatsFilter.hpp]: Yes.  Intentional.  The 'c' term is already included in the `kurtosis()` calculation.  I don't understand all the math.  `M2` is the numerator of the variance, so dividing by the square of the variance gets you two `c - 1` terms in the numerator.  One of those cancels with the `c - 1` that's in the denominator that comes up in the adjustment from the population statistic to the sample statistic.  Strangely, this finance site has some of the nicest and clearest formula: https://www.macroption.com/kurtosis-formula/

I've verified the calculations against google sheets and that against other websites (which is how I figured out that what people are selling as "kurtosis" is actually "sample excess kurtosis", so I'm pretty confident that they're correct at this point.

**Comment by abellgithub** [comment]
I'll add a case for the obvious zero value of standard deviation.

**Comment by abellgithub** [comment]
@chambbj : Think I've addressed things.

**Comment by hobu** [comment]
The bug title should be more descriptive about what was fixed here.

**Comment by abellgithub** [comment]
It references the ticket that describes the problem.

**Pull request merged**